modules/index/src/test/java/org/apache/ignite/internal/index/IndexBuildControllerTest.java

```diff
@@ -0,0 +1,280 @@
+                List.of(COLUMN_NAME)
+        );
+    }
+
+    @AfterEach
+    void tearDown() throws Exception {
+        IgniteUtils.stopAll(catalogManager);
+    }
+
+    @Test
+    void testStartBuildIndexesOnIndexCreate() {
+        newIndexBuildController();
+
+        setPrimaryReplicaForOneSecond(10, NODE_NAME, clock.now());
+
+        clearInvocations(indexBuilder);
+
+        createIndex(INDEX_NAME);
+
+        verify(indexBuilder).startBuildIndex(eq(tableId()), eq(10), eq(indexId(INDEX_NAME)), any(), any(), eq(localNode));
+    }
+
+    @Test
+    void testStartBuildIndexesOnPrimaryReplicaElected() {
+        newIndexBuildController();
+
+        createIndex(INDEX_NAME);
+
+        setPrimaryReplicaForOneSecond(10, NODE_NAME, clock.now());
+
+        verify(indexBuilder).startBuildIndex(eq(tableId()), eq(10), eq(indexId(INDEX_NAME)), any(), any(), eq(localNode));
+        verify(indexBuilder).startBuildIndex(eq(tableId()), eq(10), eq(indexId(TABLE_NAME + "_PK")), any(), any(), eq(localNode));
+    }
+
+    @Test
+    void testStopBuildIndexesOnIndexDrop() {
+        newIndexBuildController();
+
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        dropIndex(INDEX_NAME);
+
+        verify(indexBuilder).stopBuildingIndexes(indexId);
+    }
+
+    @Test
+    void testStopBuildIndexesOnChangePrimaryReplica() {
+        newIndexBuildController();
+
+        setPrimaryReplicaForOneSecond(10, NODE_NAME, clock.now());
+        setPrimaryReplicaForOneSecond(10, NODE_NAME + "_other", clock.now());
+
+        verify(indexBuilder).stopBuildingIndexes(tableId(), 10);
+    }
+
+    private void createIndex(String indexName) {
+        createHashIndex(catalogManager, DEFAULT_SCHEMA_NAME, TABLE_NAME, indexName, List.of(COLUMN_NAME), false);
+    }
+
+    private void dropIndex(String indexName) {
+        TableTestUtils.dropIndex(catalogManager, DEFAULT_SCHEMA_NAME, indexName);
+    }
+
+    private void setPrimaryReplicaForOneSecond(int partitionId, String leaseholder, HybridTimestamp startTime) {
+        CompletableFuture<ReplicaMeta> replicaMetaFuture = completedFuture(replicaMetaForOneSecond(leaseholder, startTime));
+
+        assertThat(placementDriver.setPrimaryReplicaMeta(0, replicaId(partitionId), replicaMetaFuture), willCompleteSuccessfully());
+    }
+
+    private int tableId() {
+        return getTableIdStrict(catalogManager, TABLE_NAME, clock.nowLong());
+    }
+
+    private int indexId(String indexName) {
+        return getIndexIdStrict(catalogManager, indexName, clock.nowLong());
+    }
+
+    private TablePartitionId replicaId(int partitionId) {
+        return new TablePartitionId(tableId(), partitionId);
+    }
+
+    private static TestReplicaMeta replicaMetaForOneSecond(String leaseholder, HybridTimestamp startTime) {
+        HybridTimestamp expirationTime = new HybridTimestamp(startTime.getPhysical() + 1_000, startTime.getLogical());
```

Comment:
There's a `org.apache.ignite.internal.hlc.HybridTimestamp#addPhysicalTime` that can be used for it.

Reply:
fix it